### PR TITLE
Prevent ClasspathResourceDirectoryReader from accessing closed NativeImageResourceFileSystem

### LINKED ISSUE #30344: JDBCRepositorySQLLoader  native-image之后  loadFromDirectory方法,虽然进行了文件系统的创建,如果存在抛出java.nio.file.FileSystemAlreadyExistsException异常,没有考虑
sharding版本就是当前master的代码
错误堆栈照片:
4441709173947_ pic
错误堆栈:
Caused by: java.nio.file.FileSystemAlreadyExistsException: null
 at org.graalvm.nativeimage.builder/com.oracle.svm.core.jdk.resources.NativeImageResourceFileSystemProvider.newFileSystem(NativeImageResourceFileSystemProvider.java:115) ~[ai:na]
 at java.base@21.0.2/java.nio.file.FileSystems.newFileSystem(FileSystems.java:339) ~[na:na]
 at java.base@21.0.2/java.nio.file.FileSystems.newFileSystem(FileSystems.java:288) ~[na:na]
 at org.apache.shardingsphere.mode.repository.standalone.jdbc.sql.JDBCRepositorySQLLoader.loadFromDirectory(JDBCRepositorySQLLoader.java:116) ~[na:na]
 at org.apache.shardingsphere.mode.repository.standalone.jdbc.sql.JDBCRepositorySQLLoader.load(JDBCRepositorySQLLoader.java:81) ~[na:na]
 at org.apache.shardingsphere.mode.repository.standalone.jdbc.JDBCRepository.init(JDBCRepository.java:59) ~[ai:5.4.2-SNAPSHOT]
 at org.apache.shardingsphere.infra.spi.type.typed.TypedSPILoader.findDefaultService(TypedSPILoader.java:73) ~[na:na]
 at org.apache.shardingsphere.infra.spi.type.typed.TypedSPILoader.findService(TypedSPILoader.java:57) ~[na:na]
 at org.apache.shardingsphere.infra.spi.type.typed.TypedSPILoader.getService(TypedSPILoader.java:110) ~[na:na]
 at org.apache.shardingsphere.mode.manager.standalone.NewStandaloneContextManagerBuilder.build(NewStandaloneContextManagerBuilder.java:49) ~[ai:5.4.2-SNAPSHOT]
 at org.apache.shardingsphere.driver.jdbc.core.datasource.ShardingSphereDataSource.createContextManager(ShardingSphereDataSource.java:78) ~[ai:5.4.2-SNAPSHOT]
 at org.apache.shardingsphere.driver.jdbc.core.datasource.ShardingSphereDataSource.<init>(ShardingSphereDataSource.java:66) ~[ai:5.4.2-SNAPSHOT]
 at org.apache.shardingsphere.driver.api.ShardingSphereDataSourceFactory.createDataSource(ShardingSphereDataSourceFactory.java:95) ~[ai:5.4.2-SNAPSHOT]
 at org.apache.shardingsphere.driver.api.yaml.YamlShardingSphereDataSourceFactory.createDataSource(YamlShardingSphereDataSourceFactory.java:135) ~[ai:5.4.2-SNAPSHOT]
 at org.apache.shardingsphere.driver.api.yaml.YamlShardingSphereDataSourceFactory.createDataSource(YamlShardingSphereDataSourceFactory.java:70) ~[ai:5.4.2-SNAPSHOT]
 at org.apache.shardingsphere.driver.jdbc.core.driver.DriverDataSourceCache.createDataSource(DriverDataSourceCache.java:55) ~[ai:5.4.2-SNAPSHOT]
 at org.apache.shardingsphere.driver.jdbc.core.driver.DriverDataSourceCache.lambda$get$0(DriverDataSourceCache.java:48) ~[ai:5.4.2-SNAPSHOT]
 at java.base@21.0.2/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708) ~[ai:na]
 at org.apache.shardingsphere.driver.jdbc.core.driver.DriverDataSourceCache.get(DriverDataSourceCache.java:48) ~[ai:5.4.2-SNAPSHOT]
 at org.apache.shardingsphere.driver.ShardingSphereDriver.connect(ShardingSphereDriver.java:53) ~[ai:5.4.2-SNAPSHOT]
 at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:121) ~[ai:na]
 at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:359) ~[ai:na]
 at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:201) ~[ai:na]
 at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:470) ~[ai:na]
 at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:561) ~[ai:na]
 ... 103 common frames omitted

这是我自己clone以下master的代码,进行的判断,它是成功的
4471709174079_ pic
成功的打印信息:
2024-02-29T10:56:24.891+08:00  INFO 8559 --- [               ] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Starting...
2024-02-29T10:56:24.908+08:00  INFO 8559 --- [               ] com.zaxxer.hikari.HikariDataSource       : HikariPool-2 - Starting...
2024-02-29T10:56:24.914+08:00  INFO 8559 --- [               ] com.zaxxer.hikari.pool.HikariPool        : HikariPool-2 - Added connection conn0: url=jdbc:h2:mem:config user=SA
2024-02-29T10:56:24.914+08:00  INFO 8559 --- [               ] com.zaxxer.hikari.HikariDataSource       : HikariPool-2 - Start completed.
2024-02-29T10:56:24.920+08:00  INFO 8559 --- [               ] com.alibaba.druid.pool.DruidDataSource   : {dataSource-1} inited
2024-02-29T10:56:33.231+08:00  INFO 8559 --- [               ] com.alibaba.druid.pool.DruidDataSource   : {dataSource-2} inited
2024-02-29T10:56:39.004+08:00  INFO 8559 --- [               ] com.alibaba.druid.pool.DruidDataSource   : {dataSource-3} inited
2024-02-29T10:56:41.545+08:00  INFO 8559 --- [               ] com.zaxxer.hikari.pool.PoolBase          : HikariPool-1 - Driver does not support get/set network timeout for connections. (getNetworkTimeout)
2024-02-29T10:56:41.545+08:00  INFO 8559 --- [               ] com.zaxxer.hikari.pool.HikariPool        : HikariPool-1 - Added connection org.apache.shardingsphere.driver.jdbc.core.connection.ShardingSphereConnection@59ae1330
2024-02-29T10:56:41.545+08:00  INFO 8559 --- [               ] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Start completed.
2024-02-29T10:56:41.545+08:00 DEBUG 8559 --- [               ] c.h.m.product.ProductMapper.getList      : ==>  Preparing: select id,`number`,price,original_price,`desc`,create_time from product
2024-02-29T10:56:41.591+08:00 DEBUG 8559 --- [               ] c.h.m.product.ProductMapper.getList      : ==> Parameters: 
2024-02-29T10:56:41.592+08:00  INFO 8559 --- [               ] ShardingSphere-SQL                       : Logic SQL: select id,`number`,price,original_price,`desc`,create_time from product
2024-02-29T10:56:41.592+08:00  INFO 8559 --- [               ] ShardingSphere-SQL                       : Actual SQL: ai-slave1 ::: select id,`number`,price,original_price,`desc`,create_time from product
2024-02-29T10:56:41.650+08:00 DEBUG 8559 --- [               ] c.h.m.product.ProductMapper.getList      : <==      Total: 3

### LINKED ISSUE #30345: As of merge #30321, integration tests fail on GraalVM Native Image
## Bug Report

**For English only**, other languages will not accept.

Before report a bug, make sure you have:

- Searched open and closed [GitHub issues](https://github.com/apache/shardingsphere/issues).
- Read documentation: [ShardingSphere Doc](https://shardingsphere.apache.org/document/current/en/overview).

Please pay attention on issues you submitted, because we maybe need more details. 
If no response anymore and we cannot reproduce it on current information, we will **close it**.

Please answer these questions before submitting your issue. Thanks!

### Which version of ShardingSphere did you use?

- https://github.com/apache/shardingsphere/commit/ce4cbadb39b4160e54852a6fe8066133755780c3

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?

- `ShardingSphere-JDBC`

### Expected behavior
- `./mvnw -PnativeTestInShardingSphere -T1C -e clean test` executes normally.
### Actual behavior

- As of merge #30321, integration tests fail on GraalVM Native Image.
```shell
[ERROR] 2024-02-29 11:35:30.551 [main] com.zaxxer.hikari.pool.HikariPool - HikariPool-19 - Exception during pool initialization.
java.nio.file.ClosedFileSystemException: null
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.jdk.resources.NativeImageResourceFileSystem.ensureOpen(NativeImageResourceFileSystem.java:144)
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.jdk.resources.NativeImageResourceFileSystem.getFileAttributes(NativeImageResourceFileSystem.java:303)
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.jdk.resources.NativeImageResourcePath.getAttributes(NativeImageResourcePath.java:510)
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.jdk.resources.NativeImageResourceFileSystemProvider.readAttributes(NativeImageResourceFileSystemProvider.java:236)
        at java.base@21.0.2/java.nio.file.Files.readAttributes(Files.java:1853)
        at java.base@21.0.2/java.nio.file.FileTreeWalker.getAttributes(FileTreeWalker.java:220)
        at java.base@21.0.2/java.nio.file.FileTreeWalker.visit(FileTreeWalker.java:277)
        at java.base@21.0.2/java.nio.file.FileTreeWalker.next(FileTreeWalker.java:374)
        at java.base@21.0.2/java.nio.file.FileTreeIterator.fetchNextIfNeeded(FileTreeIterator.java:83)
        at java.base@21.0.2/java.nio.file.FileTreeIterator.hasNext(FileTreeIterator.java:103)
        at java.base@21.0.2/java.util.Spliterators$IteratorSpliterator.tryAdvance(Spliterators.java:1949)
        at java.base@21.0.2/java.util.stream.StreamSpliterators$WrappingSpliterator.lambda$initPartialTraversalState$0(StreamSpliterators.java:292)
        at java.base@21.0.2/java.util.stream.StreamSpliterators$AbstractWrappingSpliterator.fillBuffer(StreamSpliterators.java:206)
        at java.base@21.0.2/java.util.stream.StreamSpliterators$AbstractWrappingSpliterator.doAdvance(StreamSpliterators.java:161)
        at java.base@21.0.2/java.util.stream.StreamSpliterators$WrappingSpliterator.tryAdvance(StreamSpliterators.java:298)
        at java.base@21.0.2/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:280)
        at java.base@21.0.2/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1685)
        at java.base@21.0.2/java.util.stream.StreamSpliterators$WrappingSpliterator.lambda$initPartialTraversalState$0(StreamSpliterators.java:292)
        at java.base@21.0.2/java.util.stream.StreamSpliterators$AbstractWrappingSpliterator.fillBuffer(StreamSpliterators.java:206)
        at java.base@21.0.2/java.util.stream.StreamSpliterators$AbstractWrappingSpliterator.doAdvance(StreamSpliterators.java:161)
        at java.base@21.0.2/java.util.stream.StreamSpliterators$WrappingSpliterator.tryAdvance(StreamSpliterators.java:298)
        at java.base@21.0.2/java.util.Spliterators$1Adapter.hasNext(Spliterators.java:681)
        at org.apache.shardingsphere.mode.repository.standalone.jdbc.sql.JDBCRepositorySQLLoader.load(JDBCRepositorySQLLoader.java:54)
        at org.apache.shardingsphere.mode.repository.standalone.jdbc.JDBCRepository.init(JDBCRepository.java:59)
        at org.apache.shardingsphere.infra.spi.type.typed.TypedSPILoader.findService(TypedSPILoader.java:61)
        at org.apache.shardingsphere.infra.spi.type.typed.TypedSPILoader.getService(TypedSPILoader.java:110)
        at org.apache.shardingsphere.mode.manager.standalone.NewStandaloneContextManagerBuilder.build(NewStandaloneContextManagerBuilder.java:49)
        at org.apache.shardingsphere.driver.jdbc.core.datasource.ShardingSphereDataSource.createContextManager(ShardingSphereDataSource.java:79)
        at org.apache.shardingsphere.driver.jdbc.core.datasource.ShardingSphereDataSource.<init>(ShardingSphereDataSource.java:67)
        at org.apache.shardingsphere.driver.api.ShardingSphereDataSourceFactory.createDataSource(ShardingSphereDataSourceFactory.java:95)
        at org.apache.shardingsphere.driver.api.yaml.YamlShardingSphereDataSourceFactory.createDataSource(YamlShardingSphereDataSourceFactory.java:135)
        at org.apache.shardingsphere.driver.api.yaml.YamlShardingSphereDataSourceFactory.createDataSource(YamlShardingSphereDataSourceFactory.java:70)
        at org.apache.shardingsphere.driver.jdbc.core.driver.DriverDataSourceCache.createDataSource(DriverDataSourceCache.java:55)
        at org.apache.shardingsphere.driver.jdbc.core.driver.DriverDataSourceCache.lambda$get$0(DriverDataSourceCache.java:48)
        at java.base@21.0.2/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708)
        at org.apache.shardingsphere.driver.jdbc.core.driver.DriverDataSourceCache.get(DriverDataSourceCache.java:48)
        at org.apache.shardingsphere.driver.ShardingSphereDriver.connect(ShardingSphereDriver.java:55)
        at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:121)
        at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:364)
        at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:206)
        at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:476)
        at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:561)
        at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:115)
        at com.zaxxer.hikari.HikariDataSource.<init>(HikariDataSource.java:81)
        at org.apache.shardingsphere.test.natived.jdbc.transactions.xa.NarayanaTest.assertShardingInNarayanaTransactions(NarayanaTest.java:40)
        at java.base@21.0.2/java.lang.reflect.Method.invoke(Method.java:580)
        at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:728)
        at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
        at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:156)
        at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:147)
        at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:86)
        at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(InterceptingExecutableInvoker.java:103)
        at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
        at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:92)
        at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:86)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:218)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:214)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:139)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:69)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
        at java.base@21.0.2/java.util.ArrayList.forEach(ArrayList.java:1596)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
        at java.base@21.0.2/java.util.ArrayList.forEach(ArrayList.java:1596)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
        at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
        at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:198)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:169)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:93)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:58)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:141)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:57)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:103)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:94)
        at org.junit.platform.launcher.core.DelegatingLauncher.execute(DelegatingLauncher.java:52)
        at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:70)
        at org.graalvm.junit.platform.NativeImageJUnitLauncher.execute(NativeImageJUnitLauncher.java:74)
        at org.graalvm.junit.platform.NativeImageJUnitLauncher.main(NativeImageJUnitLauncher.java:129)
        at java.base@21.0.2/java.lang.invoke.LambdaForm$DMH/sa346b79c.invokeStaticInit(LambdaForm$DMH)

```

### Reason analyze (If you can)

- Operations on URLs on `com.oracle.svm.core.jdk.resources.NativeImageResourceFileSystem` continue after closing `com.oracle.svm.core.jdk.resources.NativeImageResourceFileSystem`. See https://github.com/apache/shardingsphere/blob/ce4cbadb39b4160e54852a6fe8066133755780c3/mode/type/standalone/repository/provider/jdbc/src/main/java/org/apache/shardingsphere/mode/repository/standalone/jdbc/sql/JDBCRepositorySQLLoader.java .

```java
for (String each : resourceNameIterable) {
                if (!each.endsWith(FILE_EXTENSION)) {
                    continue;
                }
                JDBCRepositorySQL provider = XML_MAPPER.readValue(JDBCRepositorySQLLoader.class.getClassLoader().getResourceAsStream(each), JDBCRepositorySQL.class);
                if (provider.isDefault()) {
                    result = provider;
                }
                if (Objects.equals(provider.getType(), type)) {
                    result = provider;
                    break;
                }
            }
```

### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.

```shell
sdk use java 21.0.2-graalce
git clone git@github.com:apache/shardingsphere.git
cd ./shardingsphere/

## Use `./mvnw -PgenerateMetadata -DskipNativeTests -e -T1C clean test native:metadata-copy` and manually adjust the JSON entries

./mvnw -PnativeTestInShardingSphere -T1C -e clean test
```

### PR DESCRIPTION
**Pull request by linghengqian**
Fixes #30345.

Changes proposed in this pull request:
  - Prevent ClasspathResourceDirectoryReader from accessing closed NativeImageResourceFileSystem . Call @iamhucong to help review.
  - Check whether NativeImageResourceFileSystem has been created by a third-party dependency. Fixes #30344 .

---

Before committing this PR, I'm sure that I have checked the following options:
- [x] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [x] I have self-reviewed the commit code.
- [x] I have (or in comment I request) added corresponding labels for the pull request.
- [x] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [ ] I have made corresponding changes to the documentation.
- [ ] I have added corresponding unit tests for my changes.

### CODE REVIEW

**Comment by iamhucong** [comment]
@linghengqian Hi, I believe Stream can lazily consume resources. Are there any potential issues that I haven't considered?

**Comment by linghengqian** [comment]
> @linghengqian Hi, I believe Stream can lazily consume resources. Are there any potential issues that I haven't considered?

- The Steam obtained by the `ClasspathResourceDirectoryReader.read()` function is not read immediately, which leads to a problem with the `ClasspathResourceDirectoryReader.readDirectoryInFileSystem()` function. 
- `com.oracle.svm.core.jdk.resources.NativeImageResourceFileSystem` is opened and then closed, which causes the Stream instance returned by the `ClasspathResourceDirectoryReader.read()` function of the `JDBCRepositorySQLLoader.load()` operation to try to access a non-existent File System reads the contents of the Stream.

**Comment by iamhucong** [comment]
> > @linghengqian Hi, I believe Stream can lazily consume resources. Are there any potential issues that I haven't considered?
> 
> * The Steam obtained by the `ClasspathResourceDirectoryReader.read()` function is not read immediately, which leads to a problem with the `ClasspathResourceDirectoryReader.readDirectoryInFileSystem()` function.
> * `com.oracle.svm.core.jdk.resources.NativeImageResourceFileSystem` is opened and then closed, which causes the Stream instance returned by the `ClasspathResourceDirectoryReader.read()` function of the `JDBCRepositorySQLLoader.load()` operation to try to access a non-existent File System reads the contents of the Stream.
> * Stream is still read lazily on Jar/`jdk.nio.zipfs.ZipFileSystem`. That won't change with this PR.

```java
FileSystem resourceFileSystem = FileSystems.newFileSystem(directoryUrl.toURI(), Collections.emptyMap())
return loadFromDirectory(directoryUrl).onClose(resourceFileSystem::close());
```
Could this code resolve those problems?

**Comment by linghengqian** [comment]
> ```java
> FileSystem resourceFileSystem = FileSystems.newFileSystem(directoryUrl.toURI(), Collections.emptyMap())
> return loadFromDirectory(directoryUrl).onClose(resourceFileSystem::close());
> ```
> 
> Could this code resolve those problems?

- I don't quite understand where your logic should go. Stream is returned to `org.apache.shardingsphere.mode.repository.standalone.jdbc.sql.JDBCRepositorySQLLoader` normally. Once `JDBCRepositorySQLLoader` starts traversing `Iterable<String> resourceNameIterable`, it will start to access the closed File System. This has moved away from the logic of `org.apache.shardingsphere.infra.util.directory.ClasspathResourceDirectoryReader`.
- See https://github.com/apache/shardingsphere/blob/f356f3228fde7e37024188185b991310880940f4/mode/type/standalone/repository/provider/jdbc/src/main/java/org/apache/shardingsphere/mode/repository/standalone/jdbc/sql/JDBCRepositorySQLLoader.java .

```java
  try (Stream<String> resourceNameStream = ClasspathResourceDirectoryReader.read(JDBCRepositorySQLLoader.class.getClassLoader(), ROOT_DIRECTORY)) {
             Iterable<String> resourceNameIterable = resourceNameStream::iterator;
             for (String each : resourceNameIterable) {
```

**Comment by iamhucong** [comment]
> > ```java
> > FileSystem resourceFileSystem = FileSystems.newFileSystem(directoryUrl.toURI(), Collections.emptyMap())
> > return loadFromDirectory(directoryUrl).onClose(resourceFileSystem::close());
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Could this code resolve those problems?
> 
> * I don't quite understand where your logic should go. Stream is returned to `org.apache.shardingsphere.mode.repository.standalone.jdbc.sql.JDBCRepositorySQLLoader` normally. Once `JDBCRepositorySQLLoader` starts traversing `Iterable<String> resourceNameIterable`, it will start to access the closed File System. This has moved away from the logic of `org.apache.shardingsphere.infra.util.directory.ClasspathResourceDirectoryReader`.
> * See https://github.com/apache/shardingsphere/blob/f356f3228fde7e37024188185b991310880940f4/mode/type/standalone/repository/provider/jdbc/src/main/java/org/apache/shardingsphere/mode/repository/standalone/jdbc/sql/JDBCRepositorySQLLoader.java .
> 
> ```java
>   try (Stream<String> resourceNameStream = ClasspathResourceDirectoryReader.read(JDBCRepositorySQLLoader.class.getClassLoader(), ROOT_DIRECTORY)) {
>              Iterable<String> resourceNameIterable = resourceNameStream::iterator;
>              for (String each : resourceNameIterable) {
> ```

The `resourceFileSystem.close()` method will be called only when the resourceNameStream is closed, therefore the resourceFileSystem will be closed after the resourceNameStream has been used.

**Comment by linghengqian** [comment]
> The `resourceFileSystem.close()` method will be called only when the resourceNameStream is closed, therefore the resourceFileSystem will be closed after the resourceNameStream has been used.

- I chose this approach and made a new commit.

> Alternatively, is it possible to consider not closing the resource file system? It seems that the close() method only closes some input/output streams, but I'm not very familiar with GraalVM.

- According to https://github.com/oracle/graal/issues/7682 , the behavior of `NativeImageResourceFileSystem` is as consistent as possible with `jdk.nio.zipfs.ZipFileSystem`/Jar. Therefore, affected by https://bugs.openjdk.org/browse/JDK-8175243, developers are always required to manually switch File System `on` and `off`. 
- In the Context instance of the Spring Framework, `NativeImageResourceFileSystem` will be opened for a long time during the life cycle of the Spring Context, so the PR needs to discuss two situations.
- Therefore I don't think PR can open `NativeImageResourceFileSystem` for a long time, which may conflict with the behavior of third-party dependencies.

**Review by iamhucong — CHANGES_REQUESTED** [review]

File: infra/util/src/main/java/org/apache/shardingsphere/infra/util/directory/ClasspathResourceDirectoryReader.java [hunk 1]
```diff
@@ -162,24 +159,32 @@ private static JarFile getJarFile(final URL url) {
...
      * @param directoryUrl directory url
-     * @return stream of resource name
+     * @return list of resource name
      */
     @SneakyThrows({IOException.class, URISyntaxException.class})
-    private static Stream<String> readDirectoryInFileSystem(final URL directoryUrl) {
+    private static List<String> readDirectoryInFileSystem(final URL directoryUrl) {
         if ("resource".equals(directoryUrl.getProtocol())) {
-            try (FileSystem ignored = FileSystems.newFileSystem(URI.create("resource:/"), Collections.emptyMap())) {
+            try (FileSystem ignored = FileSystems.getFileSystem(directoryUrl.toURI())) {
```
[the remark this answers is not in the document]
Comment (linghengqian) [reply infra/util/src/main/java/org/apache/shardingsphere/infra/util/directory/ClasspathResourceDirectoryReader.java]: - According to my understanding of #30344, if `FileSystems.getFileSystem()` does not throw `java.nio.file.FileSystemNotFoundException`, it means that `com.oracle.svm.core.jdk.resources.NativeImageResourceFileSystem` is already in a third-party dependency  life cycle, and additional judgments here are meaningless. 
- So I made my third commit. Call @huyu-tom .What do you think of this? This still doesn't break NativeTest.

**Review by terrymanu — APPROVED** [review]

**Pull request merged**